File: apps/ai/app/services/embeddings.py
import math

from openai import OpenAI

from app.config import get_settings
from app.constants import EMBEDDING_MODEL

_client: OpenAI | None = None


def _get_openai() -> OpenAI:
    global _client
    if _client is None:
        settings = get_settings()
        _client = OpenAI(api_key=settings.openai_api_key)
    return _client


def embed_query(text: str) -> list[float]:
    openai = _get_openai()
    response = openai.embeddings.create(model=EMBEDDING_MODEL, input=[text])
    sorted_data = sorted(response.data, key=lambda item: item.index)
    return sorted_data[0].embedding


def embedding_to_vector(embedding: list[float]) -> str:
    return f"[{','.join(str(value) for value in embedding)}]"


def vector_from_string(raw: str) -> list[float]:
    """Inverse of embedding_to_vector — parses a pgvector column's textual
    form ("[0.01,0.02,...]", PostgREST's serialization of `vector`) back
    into floats. Used by the coverage agent's grounding-score node to reuse
    a chunk's already-computed ingest-time embedding instead of re-embedding
    its content (ADR 009: "no new API call" for the chunk side).
    """
    return [float(v) for v in raw.strip("[]").split(",")]


def cosine_similarity(a: list[float], b: list[float]) -> float:
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(y * y for y in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)

File: apps/web/lib/embeddings.ts
import OpenAI from "openai";

const EMBEDDING_MODEL = "text-embedding-3-small";
const BATCH_SIZE = 100;

let client: OpenAI | null = null;

function getOpenAI(): OpenAI {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    throw new Error("OPENAI_API_KEY is not configured.");
  }
  if (!client) {
    client = new OpenAI({ apiKey });
  }
  return client;
}

export async function embedTexts(texts: string[]): Promise<number[][]> {
  if (texts.length === 0) return [];

  const openai = getOpenAI();
  const embeddings: number[][] = [];

  for (let i = 0; i < texts.length; i += BATCH_SIZE) {
    const batch = texts.slice(i, i + BATCH_SIZE);
    const response = await openai.embeddings.create({
      model: EMBEDDING_MODEL,
      input: batch,
    });

    const sorted = [...response.data].sort((a, b) => a.index - b.index);
    for (const item of sorted) {
      embeddings.push(item.embedding);
    }
  }

  return embeddings;
}

export async function embedQuery(text: string): Promise<number[]> {
  const [embedding] = await embedTexts([text]);
  return embedding;
}

export function embeddingToVector(embedding: number[]): string {
  return `[${embedding.join(",")}]`;
}
